File: apps/web/src/api/swap.ts
/**
 * PP ↔ Token swap API client.
 */
import apiClient from '@/lib/api';

// ── Types ────────────────────────────────────────────────────────────────────

export interface SwapQuote {
  pp_amount: number;
  fee_pp: number;
  fee_percent: number;
  net_pp: number;
  token_amount: number;
  rate: number;
  direction: string;
}

export interface SwapResult {
  swap_id: string;
  pp_amount: number;
  fee_pp: number;
  token_amount: number;
  direction: string;
  rate: number;
  status: string;
  created_at: string | null;
}

export interface SwapHistoryItem {
  id: string;
  amount: number;
  transaction_type: string;
  reason: string;
  created_at: string | null;
}

export interface SwapHistory {
  items: SwapHistoryItem[];
  total: number;
  page: number;
  page_size: number;
  total_pages: number;
}

// ── API calls ────────────────────────────────────────────────────────────────

export const swapAPI = {
  /** Get a preview quote for a swap */
  getQuote: async (ppAmount: number, direction: string = 'pp_to_token'): Promise<SwapQuote> => {
    const { data } = await apiClient.get('/swap/quote', {
      params: { pp_amount: ppAmount, direction },
    });
    return data;
  },

  /** Execute a swap */
  executeSwap: async (
    ppAmount: number,
    direction: string = 'pp_to_token',
    quoteId?: string,
  ): Promise<SwapResult> => {
    const { data } = await apiClient.post('/swap/execute', {
      pp_amount: ppAmount,
      direction,
      quote_id: quoteId,
    });
    return data;
  },

  /** Get swap transaction history */
  getHistory: async (page: number = 1, pageSize: number = 20): Promise<SwapHistory> => {
    const { data } = await apiClient.get('/swap/history', {
      params: { page, page_size: pageSize },
    });
    return data;
  },
};

File: backend/app/schemas/swap.py
"""PP ↔ Token swap schemas."""
from pydantic import BaseModel, Field, ConfigDict
from typing import Optional, List
import uuid


class SwapQuoteRequest(BaseModel):
    """Request a swap quote."""
    pp_amount: float = Field(..., gt=0, description="Amount of PP to swap")
    direction: str = Field(
        "pp_to_token",
        description="Swap direction: 'pp_to_token' or 'token_to_pp'",
        pattern="^(pp_to_token|token_to_pp)$",
    )


class SwapQuoteResponse(BaseModel):
    """Preview of a swap without executing."""
    pp_amount: float
    fee_pp: float
    fee_percent: float
    net_pp: float
    token_amount: float
    rate: float
    direction: str


class SwapExecuteRequest(BaseModel):
    """Execute a swap."""
    pp_amount: float = Field(..., gt=0, description="Amount of PP to swap")
    direction: str = Field(
        "pp_to_token",
        description="Swap direction",
        pattern="^(pp_to_token|token_to_pp)$",
    )
    quote_id: Optional[str] = Field(None, description="Optional quote ID for idempotency")


class SwapExecuteResponse(BaseModel):
    """Result of an executed swap."""
    model_config = ConfigDict(from_attributes=True)

    swap_id: str
    pp_amount: float
    fee_pp: float
    token_amount: float
    direction: str
    rate: float
    status: str
    created_at: Optional[str] = None


class SwapHistoryItem(BaseModel):
    """A single swap transaction in history."""
    id: str
    amount: float
    transaction_type: str
    reason: str
    created_at: Optional[str] = None


class SwapHistoryResponse(BaseModel):
    """Paginated swap history."""
    items: List[SwapHistoryItem]
    total: int
    page: int
    page_size: int
    total_pages: int
